book/sorting.md

```diff
@@ -0,0 +1,292 @@
+# Sorting
+
+Nushell offers many ways of sorting data, and which method you reach for will depend on the problem and what kind of data you're working with. Let's take a look at some of the ways you might wish to sort data.
+
+## Basic sorting
+
+### Lists
+
+Sorting a basic list works exactly how you might expect:
+
+```nu
+> [9 3 8 1 4 6] | sort
+╭───┬───╮
+│ 0 │ 1 │
+│ 1 │ 3 │
+│ 2 │ 4 │
+│ 3 │ 6 │
+│ 4 │ 8 │
+│ 5 │ 9 │
+╰───┴───╯
+```
+
+However, things get a bit more complex when you start combining types. For example, let's see what happens when we have a list with numbers _and_ strings:
+
+```nu
+> ["hello" 4 9 2 1 "foobar" 8 6] | sort
+╭───┬────────╮
+│ 0 │      1 │
+│ 1 │      2 │
+│ 2 │      4 │
+│ 3 │      6 │
+│ 4 │      8 │
+│ 5 │      9 │
+│ 6 │ foobar │
+│ 7 │ hello  │
+╰───┴────────╯
+```
+
+We can see that the numbers are sorted in order, and the strings are sorted to the end of the list, also in order. If you are coming from other programming languages, this may not be quite what you expect. In Nushell, as a general rule, **data can always be sorted without erroring**.
```

Comment:
Some part of my inner "language patent lawyer" thinks that promising this leads to some tricky rules to define a total order over all data types with challenges as soon as new types would be introduced and are messy to teach if you run into them. But for the general usability it is definitely a bonus if you don't have to worry about this.

Reply:
At least currently, there is indeed an ordering between all types. It's effectively the order of the `Value` enum's variants, with some exceptions. As I see it, the tradeoff here is between "sort which errors on mixed types" or "sort that never errors". Personally, if it would be possible for sort to error because of incompatible types, I would rather make it fully strict at that point.

I wasn't planning on specifying what order the types would be, other than that within the same version of Nushell the ordering between types is always the same. I would like to guarantee at least that `null` comes at the end, and the corresponding Nushell PR does do this. Since Nothing is kind of a special type, and you might expect it to appear a lot more than say, an integer and a string, I think it's a useful property for it to always appear at the end.

Reply:
I think it's unlikely that we will add more types (I would probably be against it). For the current set of types, a total order is totally doable. Floats are really the only issue, and could be ordered using [`total_cmp`](https://doc.rust-lang.org/std/primitive.f64.html#method.total_cmp) or sorting NaNs last like [`ordered_float`](https://docs.rs/ordered-float/4.4.0/ordered_float/struct.OrderedFloat.html).

Reply:
Oh, good point. I might look into adding a total order 🤔